tests/test_datatype.py

```diff
@@ -59,3 +61,90 @@ def test_process_job_control_options_resolution():
+        ("docker-registry.crim.ca/repo/image:latest",
+         "docker-registry.crim.ca/repo/image:latest", "docker-registry.crim.ca", "repo/image:latest"),
+        ("docker-registry.crim.ca/repo/image:1.0.0",
+         "docker-registry.crim.ca/repo/image:1.0.0", "docker-registry.crim.ca", "repo/image:1.0.0"),
+        ("quay.io/prometheus/node-exporter:v1.0.0",
+         "quay.io/prometheus/node-exporter:v1.0.0", "quay.io", "prometheus/node-exporter:v1.0.0"),
+        ("pavics/jupyterhub:1.3.0-20201211",
+         "pavics/jupyterhub:1.3.0-20201211", docker_hub, "pavics/jupyterhub:1.3.0-20201211"),
+        ("registry.gitlab.com/crim.ca/category/group/project:1.2.3",
+         "registry.gitlab.com/crim.ca/category/group/project:1.2.3",
+         "registry.gitlab.com", "crim.ca/category/group/project:1.2.3"),
+        ("https://index.docker.io/v1/repo/image:test",
+         "repo/image:test", docker_hub, "repo/image:test"),
+        ("registry.example.com/org/image-name",
+         "registry.example.com/org/image-name", "registry.example.com", "org/image-name"),
+        ("registry.example.com/org/image-name:version",
+         "registry.example.com/org/image-name:version", "registry.example.com", "org/image-name:version"),
+        ("repository/image-name:version",
+         "repository/image-name:version", docker_hub, "repository/image-name:version"),
+        ("repository/image-name",
+         "repository/image-name", docker_hub, "repository/image-name"),
+    ]
+    invalid_references = [
+        # missing repo part, not allowed local/public images
+        "debian:stretch-slim",  # valid image, but public so no reason to have auth applied for it
+        "image-name:version",
+        "image-name",
+        # not a URI repository (nowhere to send Auth token since not default DockerHub)
+        "repository/org/image-name",
+        "repository/org/image-name:version",
+    ]
+    token = str(uuid.uuid4())
+    for docker_input, docker_ref, docker_registry, docker_image in valid_references:
+        try:
+            auth = DockerAuthentication("Basic", token, docker_input)
+            assert auth.token == token, f"Testing: [{docker_input}]"
+            assert auth.registry == docker_registry, f"Testing: [{docker_input}]"
+            assert auth.image == docker_image, f"Testing: [{docker_input}]"
+            assert auth.docker == docker_ref, f"Testing: [{docker_input}]"
+            assert auth.link == docker_input, f"Testing: [{docker_input}]"
+        except (TypeError, ValueError) as exc:
+            pytest.fail(f"Unexpected failure when [{docker_input}] was expected to be valid: [{exc}]")
+    for docker_input in invalid_references:
+        try:
+            DockerAuthentication("Basic", token, docker_input)
+        except (TypeError, ValueError):
+            pass
+        else:
+            raise AssertionError(f"Testing [{docker_input}] have raised invalid format but it did not.")
```

Comment:
Not sure if error message is incomplete here ("but it did not ..?").

Reply:
-> should have raised, but it did not?

Reply:
I reword. It is like @dbyrns said.